src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -42,18 +45,31 @@ def parse_interpreter_constraints(python_setup, python_target_adaptors):
   return constraints_args
 
 
+def resolve_all_transitive_hydrated_targets(initial_transitive_hydrated_target):
```

Comment:
I think this should be equivalent to `all_targets = transitive_hydrated_target.closure`...

Reply:
That sounds very nifty, and I see now that that is available in V1. It doesn't look to be implemented in V2:

```
(Pdb) dir(transitive_hydrated_target)
['__abstractmethods__', '__add__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmul__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_cache', '_abc_negative_cache', '_abc_negative_cache_version', '_abc_registry', '_asdict', '_fields', '_make', '_replace', '_source', '_super_iter', 'copy', 'count', 'dependencies', 'index', 'make_type_error', 'root', 'type_check_error_type']

(Pdb) dir(transitive_hydrated_target.root)
['__abstractmethods__', '__add__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmul__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_cache', '_abc_negative_cache', '_abc_negative_cache_version', '_abc_registry', '_asdict', '_fields', '_make', '_replace', '_source', '_super_iter', 'adaptor', 'address', 'addresses', 'copy', 'count', 'dependencies', 'index', 'make_type_error', 'type_check_error_type']

(Pdb) dir(transitive_hydrated_target.root.adaptor)
['_ABSTRACT_FIELD', '_INHERITANCE_FIELDS', '_INTERNAL_FIELDS', '_TYPE_ALIAS_FIELD', '_UNINHERITABLE_FIELDS', '__abstractmethods__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_abc_cache', '_abc_negative_cache', '_abc_negative_cache_version', '_abc_registry', '_asdict', '_extract_inheritable_attributes', '_key', '_kwargs', 'abstract', 'address', 'create', 'default_sources_exclude_globs', 'default_sources_globs', 'dependencies', 'extends', 'field_adaptors', 'get_sources', 'is_serializable', 'is_serializable_type', 'kwargs', 'merges', 'name', 'report_validation_error', 'type_alias', 'validate', 'validate_concrete', 'validate_sources']
```

Should this PR be focused on adding a `closure` attribute to all transitive hydrated targets? Would this be defined on the `target_adaptor`, the `hydrated_target`, or the `transitive_hydrated_target`?

Reply:
Oh sorry, I mis-read slightly. If you request a `TransitiveHydratedTargets` instead of a `TransitiveHydratedTarget`, that has a `root` field which is the `TransitiveHydratedTarget`, and a `closure` field which is a flattened tuple of the transitive deps.

Reply:
Opened up https://github.com/pantsbuild/pants/issues/7726 to make this possible.

Reply:
Re: #7726: looking at the rule graph rendered by `./pants --native-engine-visualize-to=blah help`, I see:
![Screen Shot 2019-05-14 at 11 48 14 AM](https://user-images.githubusercontent.com/46740/57723807-48075200-763e-11e9-968d-d9825550bfd2.png)

Which indicates that you should be able to get TransitiveHydratedTargets using BuildFileAddresses. So:
```
transitive_hydrated_targets = yield Get(TransitiveHydratedTargets, BuildFileAddresses([address_of_the_root_target_i_care_about]))
```

Reply:
Great find! It's indeed a bit awkward, but better than defining our own recursive traversal code. I updated the issue to reflect that we now use this.